src/nibbles.rs

```diff
@@ -75,75 +139,40 @@ impl Nibbles {
             0x00
         };
 
+        let mut compact = Vec::with_capacity(hex.len() / 2 + 1);
+
         compact.push(v + if is_leaf { 0x20 } else { 0x00 });
-        for i in 0..(hex.len() / 2) {
-            compact.push((hex[i * 2] * 16) + (hex[i * 2 + 1]));
+        for hex in hex.chunks_exact(2) {
+            compact.push((hex[0] * 16) + (hex[1]));
         }
 
         compact
     }
 
     pub fn encode_raw(&self) -> (Vec<u8>, bool) {
-        let mut raw = vec![];
         let is_leaf = self.is_leaf();
-        let hex = if is_leaf {
-            &self.hex_data[0..self.hex_data.len() - 1]
-        } else {
-            &self.hex_data[0..]
-        };
-
-        for i in 0..(hex.len() / 2) {
-            raw.push((hex[i * 2] * 16) + (hex[i * 2 + 1]));
+        let mut raw = Vec::with_capacity(self.hex_data.len() / 2);
+        // if `is_leaf` then we don't care about the last nibble anyway.
+        for hex in self.hex_data.chunks_exact(2) {
+            raw.push((hex[0] * 16) + (hex[1]));
         }
 
         (raw, is_leaf)
     }
 
-    pub fn len(&self) -> usize {
-        self.hex_data.len()
-    }
-
-    pub fn is_empty(&self) -> bool {
-        self.len() == 0
-    }
-
-    pub fn at(&self, i: usize) -> usize {
-        self.hex_data[i] as usize
-    }
-
-    pub fn common_prefix(&self, other_partial: &Nibbles) -> usize {
-        let s = min(self.len(), other_partial.len());
-        let mut i = 0usize;
-        while i < s {
-            if self.at(i) != other_partial.at(i) {
-                break;
-            }
-            i += 1;
-        }
-        i
-    }
-
-    pub fn offset(&self, index: usize) -> Nibbles {
-        self.slice(index, self.hex_data.len())
-    }
-
-    pub fn slice(&self, start: usize, end: usize) -> Nibbles {
-        Nibbles::from_hex(self.hex_data[start..end].to_vec())
-    }
-
-    pub fn get_data(&self) -> &[u8] {
+    pub fn _get_data(&self) -> &[u8] {
```

Comment:
Out of curiosity: why is there an underscore prefix?

Reply:
I used it some commits ago and then stopped because it was unneeded. I deleted in 9df094f, thanks!